File: frontend/src/services/api/orchard.ts
/**
 * Orchard Privacy Protocol API Service
 *
 * API endpoints for Zcash Orchard shielded transfers.
 */

import axios from './axios';
import type {
  CombinedZcashBalance,
  EnableOrchardRequest,
  EnableOrchardResponse,
  GenerateAddressRequest,
  OrchardTransactionInfo,
  OrchardTransferRequest,
  OrchardTransferResponse,
  ScanProgress,
  ShieldedBalance,
  UnifiedAddressInfo,
} from '../../types/orchard';

/**
 * Enable Orchard for a Zcash wallet
 *
 * This derives Orchard keys and generates a unified address.
 */
export async function enableOrchard(
  request: EnableOrchardRequest
): Promise<EnableOrchardResponse> {
  const response = await axios.post<EnableOrchardResponse>(
    `/wallets/${request.wallet_id}/orchard/enable`,
    { birthday_height: request.birthday_height }
  );
  return response.data;
}

/**
 * Get all unified addresses for a wallet
 */
export async function getUnifiedAddresses(
  walletId: number
): Promise<UnifiedAddressInfo[]> {
  const response = await axios.get<UnifiedAddressInfo[]>(
    `/wallets/${walletId}/orchard/addresses`
  );
  return response.data;
}

/**
 * Generate a new unified address
 */
export async function generateUnifiedAddress(
  walletId: number,
  request: GenerateAddressRequest
): Promise<UnifiedAddressInfo> {
  const response = await axios.post<UnifiedAddressInfo>(
    `/wallets/${walletId}/orchard/addresses`,
    request
  );
  return response.data;
}

/**
 * Get shielded (Orchard) balance for a wallet
 */
export async function getShieldedBalance(
  walletId: number
): Promise<ShieldedBalance> {
  const response = await axios.get<ShieldedBalance>(
    `/wallets/${walletId}/orchard/balance`
  );
  return response.data;
}

/**
 * Get combined balance (transparent + shielded)
 */
export async function getCombinedBalance(
  walletId: number
): Promise<CombinedZcashBalance> {
  const response = await axios.get<CombinedZcashBalance>(
    `/wallets/${walletId}/orchard/balance/combined`
  );
  return response.data;
}

/**
 * Initiate an Orchard (shielded) transfer
 */
export async function initiateOrchardTransfer(
  request: OrchardTransferRequest
): Promise<OrchardTransferResponse> {
  const response = await axios.post<OrchardTransferResponse>(
    '/transfers/orchard',
    request
  );
  return response.data;
}

/**
 * Execute a pending Orchard transfer
 */
export async function executeOrchardTransfer(
  transferId: number
): Promise<OrchardTransferResponse> {
  const response = await axios.post<OrchardTransferResponse>(
    `/transfers/orchard/${transferId}/execute`
  );
  return response.data;
}

/**
 * Get Orchard transaction history
 */
export async function getOrchardTransactions(
  walletId: number,
  limit: number = 50
): Promise<OrchardTransactionInfo[]> {
  const response = await axios.get<OrchardTransactionInfo[]>(
    `/wallets/${walletId}/orchard/transactions`,
    { params: { limit } }
  );
  return response.data;
}

/**
 * Get scan progress
 */
export async function getScanProgress(): Promise<ScanProgress> {
  const response = await axios.get<ScanProgress>('/zcash/scan/status');
  return response.data;
}

/**
 * Trigger a sync of the Orchard scanner
 */
export async function syncOrchard(): Promise<ScanProgress> {
  const response = await axios.post<ScanProgress>('/zcash/scan/sync');
  return response.data;
}

/**
 * Parse a unified address to get its components
 */
export async function parseUnifiedAddress(
  address: string
): Promise<UnifiedAddressInfo> {
  const response = await axios.get<UnifiedAddressInfo>(
    '/zcash/address/parse',
    { params: { address } }
  );
  return response.data;
}

/**
 * Validate a unified address
 */
export async function validateUnifiedAddress(
  address: string
): Promise<{ valid: boolean; error?: string }> {
  try {
    const response = await axios.get<{ valid: boolean }>(
      '/zcash/address/validate',
      { params: { address } }
    );
    return response.data;
  } catch (error: any) {
    return {
      valid: false,
      error: error.response?.data?.message || 'Invalid address',
    };
  }
}

/**
 * Estimate fee for an Orchard transfer
 */
export async function estimateOrchardFee(
  request: Omit<OrchardTransferRequest, 'wallet_id'>
): Promise<{
  fee_zatoshis: number;
  fee_zec: string;
  num_actions: number;
}> {
  const response = await axios.post<{
    fee_zatoshis: number;
    fee_zec: string;
    num_actions: number;
  }>('/transfers/orchard/estimate-fee', request);
  return response.data;
}

/**
 * Check if Orchard is enabled for a wallet
 */
export async function isOrchardEnabled(walletId: number): Promise<boolean> {
  try {
    await getShieldedBalance(walletId);
    return true;
  } catch {
    return false;
  }
}

// Export all functions as a namespace for convenient imports
const orchardApi = {
  enableOrchard,
  getUnifiedAddresses,
  generateUnifiedAddress,
  getShieldedBalance,
  getCombinedBalance,
  initiateOrchardTransfer,
  executeOrchardTransfer,
  getOrchardTransactions,
  getScanProgress,
  syncOrchard,
  parseUnifiedAddress,
  validateUnifiedAddress,
  estimateOrchardFee,
  isOrchardEnabled,
};

export default orchardApi;

File: backend/src/api/handlers/orchard.rs
//! Orchard Privacy Protocol API Handlers
//!
//! Handles API requests for Zcash Orchard shielded transfers.

use actix_web::{web, HttpResponse};
use serde::{Deserialize, Serialize};
use std::sync::Arc;

use crate::api::middleware::AuthenticatedUser;
use crate::error::{AppError, AppResult};
use crate::services::WalletService;

/// Request to enable Orchard for a wallet
#[derive(Debug, Deserialize)]
pub struct EnableOrchardRequest {
    pub birthday_height: u64,
}

/// Response after enabling Orchard
#[derive(Debug, Serialize)]
pub struct EnableOrchardResponse {
    pub unified_address: UnifiedAddressInfo,
    pub viewing_key: String,
}

/// Unified address information
#[derive(Debug, Serialize)]
pub struct UnifiedAddressInfo {
    pub address: String,
    pub has_orchard: bool,
    pub has_sapling: bool,
    pub has_transparent: bool,
    pub transparent_address: Option<String>,
    pub address_index: u32,
    pub account_index: u32,
}

/// Shielded balance response
#[derive(Debug, Serialize)]
pub struct ShieldedBalanceResponse {
    pub total_zatoshis: u64,
    pub spendable_zatoshis: u64,
    pub pending_zatoshis: u64,
    pub note_count: u32,
    pub pool: String,
}

/// Combined balance response
#[derive(Debug, Serialize)]
pub struct CombinedBalanceResponse {
    pub wallet_id: i32,
    pub address: String,
    pub transparent_balance: String,
    pub shielded_balance: Option<ShieldedBalanceResponse>,
    pub total_zec: f64,
}

/// Scan progress response
#[derive(Debug, Serialize)]
pub struct ScanProgressResponse {
    pub chain: String,
    pub scan_type: String,
    pub last_scanned_height: u64,
    pub chain_tip_height: u64,
    pub progress_percent: f64,
    pub estimated_seconds_remaining: Option<u64>,
    pub is_scanning: bool,
    pub notes_found: u64,
}

/// Orchard transfer request
#[derive(Debug, Deserialize)]
pub struct OrchardTransferRequest {
    pub wallet_id: i32,
    pub to_address: String,
    pub amount: String,
    pub amount_zatoshis: Option<u64>,
    pub memo: Option<String>,
    pub target_pool: Option<String>,
}

/// Orchard transfer response
#[derive(Debug, Serialize)]
pub struct OrchardTransferResponse {
    pub transfer_id: i64,
    pub status: String,
    pub tx_hash: Option<String>,
}

/// Enable Orchard for a Zcash wallet
pub async fn enable_orchard(
    wallet_service: web::Data<Arc<WalletService>>,
    user: AuthenticatedUser,
    path: web::Path<i32>,
    request: web::Json<EnableOrchardRequest>,
) -> AppResult<HttpResponse> {
    if user.role != "admin" {
        return Err(AppError::Forbidden("Only admin can enable Orchard".to_string()));
    }

    let wallet_id = path.into_inner();

    let (unified_address, viewing_key) = wallet_service
        .enable_orchard(wallet_id, request.birthday_height)
        .await?;

    let response = EnableOrchardResponse {
        unified_address: UnifiedAddressInfo {
            address: unified_address.address,
            has_orchard: unified_address.has_orchard,
            has_sapling: unified_address.has_sapling,
            has_transparent: unified_address.has_transparent,
            transparent_address: unified_address.transparent_address,
            address_index: unified_address.address_index,
            account_index: unified_address.account_index,
        },
        viewing_key,
    };

    Ok(HttpResponse::Ok().json(response))
}

/// Get shielded balance for a wallet
pub async fn get_shielded_balance(
    wallet_service: web::Data<Arc<WalletService>>,
    path: web::Path<i32>,
) -> AppResult<HttpResponse> {
    let wallet_id = path.into_inner();
    let balance = wallet_service.get_shielded_balance(wallet_id).await?;

    let response = ShieldedBalanceResponse {
        total_zatoshis: balance.total_zatoshis,
        spendable_zatoshis: balance.spendable_zatoshis,
        pending_zatoshis: balance.pending_zatoshis,
        note_count: balance.note_count,
        pool: format!("{:?}", balance.pool).to_lowercase(),
    };

    Ok(HttpResponse::Ok().json(response))
}

/// Get combined balance (transparent + shielded)
pub async fn get_combined_balance(
    wallet_service: web::Data<Arc<WalletService>>,
    path: web::Path<i32>,
) -> AppResult<HttpResponse> {
    let wallet_id = path.into_inner();
    let balance = wallet_service.get_combined_zcash_balance(wallet_id).await?;

    let response = CombinedBalanceResponse {
        wallet_id: balance.wallet_id,
        address: balance.address,
        transparent_balance: balance.transparent_balance,
        shielded_balance: balance.shielded_balance.map(|b| ShieldedBalanceResponse {
            total_zatoshis: b.total_zatoshis,
            spendable_zatoshis: b.spendable_zatoshis,
            pending_zatoshis: b.pending_zatoshis,
            note_count: b.note_count,
            pool: format!("{:?}", b.pool).to_lowercase(),
        }),
        total_zec: balance.total_zec,
    };

    Ok(HttpResponse::Ok().json(response))
}

/// Get scan progress
pub async fn get_scan_progress(
    wallet_service: web::Data<Arc<WalletService>>,
) -> AppResult<HttpResponse> {
    let progress = wallet_service.get_scan_progress().await?;

    let response = ScanProgressResponse {
        chain: progress.chain,
        scan_type: progress.scan_type,
        last_scanned_height: progress.last_scanned_height,
        chain_tip_height: progress.chain_tip_height,
        progress_percent: progress.progress_percent,
        estimated_seconds_remaining: progress.estimated_seconds_remaining,
        is_scanning: progress.is_scanning,
        notes_found: progress.notes_found,
    };

    Ok(HttpResponse::Ok().json(response))
}

/// Trigger Orchard sync
pub async fn sync_orchard(
    wallet_service: web::Data<Arc<WalletService>>,
    user: AuthenticatedUser,
) -> AppResult<HttpResponse> {
    if user.role != "admin" {
        return Err(AppError::Forbidden("Only admin can trigger sync".to_string()));
    }

    let progress = wallet_service.sync_orchard().await?;

    let response = ScanProgressResponse {
        chain: progress.chain,
        scan_type: progress.scan_type,
        last_scanned_height: progress.last_scanned_height,
        chain_tip_height: progress.chain_tip_height,
        progress_percent: progress.progress_percent,
        estimated_seconds_remaining: progress.estimated_seconds_remaining,
        is_scanning: progress.is_scanning,
        notes_found: progress.notes_found,
    };

    Ok(HttpResponse::Ok().json(response))
}

/// Initiate an Orchard transfer
pub async fn initiate_orchard_transfer(
    _wallet_service: web::Data<Arc<WalletService>>,
    user: AuthenticatedUser,
    request: web::Json<OrchardTransferRequest>,
) -> AppResult<HttpResponse> {
    if user.role != "admin" {
        return Err(AppError::Forbidden("Only admin can initiate transfers".to_string()));
    }

    // TODO: Implement actual transfer initiation
    // For now, return a mock response
    let response = OrchardTransferResponse {
        transfer_id: 1,
        status: "pending".to_string(),
        tx_hash: None,
    };

    tracing::info!(
        "Orchard transfer initiated: wallet={}, to={}, amount={}",
        request.wallet_id,
        request.to_address,
        request.amount
    );

    Ok(HttpResponse::Ok().json(response))
}

/// Execute a pending Orchard transfer
pub async fn execute_orchard_transfer(
    _wallet_service: web::Data<Arc<WalletService>>,
    user: AuthenticatedUser,
    path: web::Path<i64>,
) -> AppResult<HttpResponse> {
    if user.role != "admin" {
        return Err(AppError::Forbidden("Only admin can execute transfers".to_string()));
    }

    let transfer_id = path.into_inner();

    // TODO: Implement actual transfer execution
    // For now, return a mock response
    let response = OrchardTransferResponse {
        transfer_id,
        status: "submitted".to_string(),
        tx_hash: Some("mock_tx_hash_placeholder".to_string()),
    };

    tracing::info!("Orchard transfer executed: id={}", transfer_id);

    Ok(HttpResponse::Ok().json(response))
}
